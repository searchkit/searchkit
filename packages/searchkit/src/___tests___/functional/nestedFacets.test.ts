import Client, { AlgoliaMultipleQueriesQuery } from '../../'
import nock from 'nock'
import {
  NestedQueryRequest,
  NumericFilterNestedQueryRequest,
  TwoFiltersNestedQueryRequest
} from '../mocks/AlgoliaRequests'
import {
  ExampleNestedFiltersResponse,
  ExampleNestedMixedFacetResponse,
  ExampleNestedNumericFilterResponse
} from '../mocks/ElasticsearchResponses'

describe('Nested Facets, filters and results', () => {
  it('One nested facet support', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title'],
        search_attributes: ['title'],
        result_attributes: ['title'],
        facet_attributes: [
          {
            attribute: 'user.first',
            field: 'first.keyword',
            type: 'string',
            nestedPath: 'user'
          }
        ]
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1])

        expect(x.aggs).toMatchInlineSnapshot(`
          {
            "user.": {
              "aggs": {
                "user.first": {
                  "terms": {
                    "field": "user.first.keyword",
                    "size": 10,
                  },
                },
              },
              "nested": {
                "path": "user",
              },
            },
          }
        `)

        // has the nested filter applied
        expect(x.query).toMatchInlineSnapshot(`
          {
            "bool": {
              "filter": [
                {
                  "bool": {
                    "should": [
                      {
                        "nested": {
                          "inner_hits": {},
                          "path": "user",
                          "query": {
                            "bool": {
                              "should": [
                                {
                                  "term": {
                                    "user.first.keyword": "Alice",
                                  },
                                },
                              ],
                            },
                          },
                        },
                      },
                    ],
                  },
                },
              ],
              "must": {
                "match_all": {},
              },
            },
          }
        `)
        return true
      })
      .reply(200, ExampleNestedFiltersResponse)

    const response = await client.handleInstantSearchRequests(
      NestedQueryRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })

  it('one facet and two OR filters', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title'],
        search_attributes: ['title'],
        result_attributes: ['title'],
        facet_attributes: [
          {
            attribute: 'user.first',
            field: 'first.keyword',
            type: 'string',
            nestedPath: 'user'
          }
        ]
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1])

        expect(x.aggs).toMatchInlineSnapshot(`
          {
            "user.": {
              "aggs": {
                "user.first": {
                  "terms": {
                    "field": "user.first.keyword",
                    "size": 10,
                  },
                },
              },
              "nested": {
                "path": "user",
              },
            },
          }
        `)

        // has the nested filter applied
        expect(x.query).toMatchInlineSnapshot(`
          {
            "bool": {
              "filter": [
                {
                  "bool": {
                    "should": [
                      {
                        "nested": {
                          "inner_hits": {},
                          "path": "user",
                          "query": {
                            "bool": {
                              "should": [
                                {
                                  "term": {
                                    "user.first.keyword": "Alice",
                                  },
                                },
                                {
                                  "term": {
                                    "user.first.keyword": "John",
                                  },
                                },
                              ],
                            },
                          },
                        },
                      },
                    ],
                  },
                },
              ],
              "must": {
                "match_all": {},
              },
            },
          }
        `)
        return true
      })
      .reply(200, ExampleNestedFiltersResponse)

    const response = await client.handleInstantSearchRequests(
      TwoFiltersNestedQueryRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })

  it('one numeric facet', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title'],
        search_attributes: ['title'],
        result_attributes: ['title'],
        facet_attributes: [
          {
            attribute: 'user.price',
            field: 'price',
            type: 'numeric',
            nestedPath: 'user'
          }
        ]
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1])

        expect(x.aggs).toMatchInlineSnapshot(`
          {
            "user.": {
              "aggs": {
                "user.price$_entries": {
                  "terms": {
                    "field": "user.price",
                    "size": 10,
                  },
                },
                "user.price$_stats": {
                  "stats": {
                    "field": "user.price",
                  },
                },
              },
              "nested": {
                "path": "user",
              },
            },
          }
        `)

        // has the nested filter applied
        expect(x.query).toMatchInlineSnapshot(`
          {
            "bool": {
              "filter": [
                {
                  "nested": {
                    "inner_hits": {},
                    "path": "user",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "user.price": {
                                "gte": "90",
                              },
                            },
                          },
                        ],
                      },
                    },
                  },
                },
              ],
              "must": {
                "match_all": {},
              },
            },
          }
        `)
        return true
      })
      .reply(200, ExampleNestedNumericFilterResponse)

    const response = await client.handleInstantSearchRequests(
      NumericFilterNestedQueryRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })

  it('two nested facets', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title'],
        search_attributes: ['title'],
        result_attributes: ['title'],
        facet_attributes: [
          {
            attribute: 'user.price',
            field: 'price',
            type: 'numeric',
            nestedPath: 'user'
          },
          {
            attribute: 'user.first',
            field: 'first.keyword',
            type: 'string',
            nestedPath: 'user'
          }
        ]
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1])

        expect(x.aggs).toMatchInlineSnapshot(`
          {
            "user.": {
              "aggs": {
                "user.first": {
                  "terms": {
                    "field": "user.first.keyword",
                    "size": 10,
                  },
                },
                "user.price$_entries": {
                  "terms": {
                    "field": "user.price",
                    "size": 10,
                  },
                },
                "user.price$_stats": {
                  "stats": {
                    "field": "user.price",
                  },
                },
              },
              "nested": {
                "path": "user",
              },
            },
          }
        `)

        // has the nested filter applied
        expect(x.query).toMatchInlineSnapshot(`
          {
            "bool": {
              "filter": [
                {
                  "nested": {
                    "inner_hits": {},
                    "path": "user",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "user.price": {
                                "gte": "90",
                              },
                            },
                          },
                        ],
                      },
                    },
                  },
                },
              ],
              "must": {
                "match_all": {},
              },
            },
          }
        `)
        return true
      })
      .reply(200, ExampleNestedMixedFacetResponse)

    const response = await client.handleInstantSearchRequests(
      NumericFilterNestedQueryRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })
})
