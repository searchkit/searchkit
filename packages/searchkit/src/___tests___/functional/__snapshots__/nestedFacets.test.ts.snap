// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested Facets, filters and results One nested facet support 3`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.first": {
          "Alice": 1,
          "John": 2,
          "Smith": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "1",
        },
        {
          "_highlightResult": {},
          "objectID": "2",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 2,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=user.first%3AAlice&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&page=0&query=&tagFilters=",
      "processingTimeMS": 1,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.first",
            ],
          },
          "values": {
            "user.first": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.first": {
          "Alice": 1,
          "Haribo": 1,
          "Jacob": 1,
          "John": 2,
          "Smith": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "1",
        },
      ],
      "hitsPerPage": 1,
      "index": "imdb_movies",
      "nbHits": 3,
      "nbPages": 3,
      "page": 0,
      "params": "analytics=false&attributesToHighlight=&attributesToRetrieve=&attributesToSnippet=&clickAnalytics=false&facets=user.first&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&hitsPerPage=1&maxValuesPerFacet=10&page=0&query=&tagFilters=",
      "processingTimeMS": 0,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.first",
            ],
          },
          "values": {
            "user.first": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`Nested Facets, filters and results One nested facet support: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.first":{"terms":{"field":"user.first.keyword","size":10}}}}},"query":{"bool":{"filter":[{"bool":{"should":[{"nested":{"inner_hits":{},"path":"user","query":{"bool":{"should":[{"term":{"user.first.keyword":"Alice"}}]}}}}]}}],"must":{"match_all":{}}}},"size":20,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
{"index":"imdb_movies"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.first":{"terms":{"field":"user.first.keyword","size":10}}}}},"query":{"bool":{"filter":[],"must":{"match_all":{}}}},"size":1,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
"
`;

exports[`Nested Facets, filters and results one facet and two OR filters 3`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.first": {
          "Alice": 1,
          "John": 2,
          "Smith": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "1",
        },
        {
          "_highlightResult": {},
          "objectID": "2",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 2,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=user.first%3AAlice%2Cuser.first%3AJohn&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&page=0&query=&tagFilters=",
      "processingTimeMS": 1,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.first",
            ],
          },
          "values": {
            "user.first": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.first": {
          "Alice": 1,
          "Haribo": 1,
          "Jacob": 1,
          "John": 2,
          "Smith": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "1",
        },
      ],
      "hitsPerPage": 1,
      "index": "imdb_movies",
      "nbHits": 3,
      "nbPages": 3,
      "page": 0,
      "params": "analytics=false&attributesToHighlight=&attributesToRetrieve=&attributesToSnippet=&clickAnalytics=false&facets=user.first&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&hitsPerPage=1&maxValuesPerFacet=10&page=0&query=&tagFilters=",
      "processingTimeMS": 0,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.first",
            ],
          },
          "values": {
            "user.first": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`Nested Facets, filters and results one facet and two OR filters: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.first":{"terms":{"field":"user.first.keyword","size":10}}}}},"query":{"bool":{"filter":[{"bool":{"should":[{"nested":{"inner_hits":{},"path":"user","query":{"bool":{"should":[{"term":{"user.first.keyword":"Alice"}},{"term":{"user.first.keyword":"John"}}]}}}}]}}],"must":{"match_all":{}}}},"size":20,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
{"index":"imdb_movies"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.first":{"terms":{"field":"user.first.keyword","size":10}}}}},"query":{"bool":{"filter":[],"must":{"match_all":{}}}},"size":1,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
"
`;

exports[`Nested Facets, filters and results one numeric facet 3`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.price": {
          "123": 1,
          "34": 1,
        },
      },
      "facets_stats": {
        "user.price": {
          "avg": 78.5,
          "max": 123,
          "min": 34,
          "sum": 157,
        },
      },
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "4",
        },
      ],
      "hitsPerPage": 20,
      "index": "my-index-000001",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&numericFilters=user.price%3E%3D90&page=0&query=&tagFilters=",
      "processingTimeMS": 1,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.price",
            ],
          },
          "values": {
            "user.price": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`Nested Facets, filters and results one numeric facet: ES Request 1`] = `
"{"index":"my-index-000001"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.price$_stats":{"stats":{"field":"user.price"}},"user.price$_entries":{"terms":{"field":"user.price","size":10}}}}},"query":{"bool":{"filter":[{"nested":{"path":"user","inner_hits":{},"query":{"bool":{"filter":[{"range":{"user.price":{"gte":"90"}}}]}}}}],"must":{"match_all":{}}}},"size":20,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
"
`;

exports[`Nested Facets, filters and results two nested facets 3`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.first": {
          "Haribo": 1,
          "Jacob": 1,
        },
        "user.price": {
          "123": 1,
          "34": 1,
        },
      },
      "facets_stats": {
        "user.price": {
          "avg": 78.5,
          "max": 123,
          "min": 34,
          "sum": 157,
        },
      },
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "4",
        },
      ],
      "hitsPerPage": 20,
      "index": "my-index-000001",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&numericFilters=user.price%3E%3D90&page=0&query=&tagFilters=",
      "processingTimeMS": 1,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.price",
              "user.first",
            ],
          },
          "values": {
            "user.first": {
              "sortRemainingBy": "count",
            },
            "user.price": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`Nested Facets, filters and results two nested facets: ES Request 1`] = `
"{"index":"my-index-000001"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.price$_stats":{"stats":{"field":"user.price"}},"user.price$_entries":{"terms":{"field":"user.price","size":10}},"user.first":{"terms":{"field":"user.first.keyword","size":10}}}}},"query":{"bool":{"filter":[{"nested":{"path":"user","inner_hits":{},"query":{"bool":{"filter":[{"range":{"user.price":{"gte":"90"}}}]}}}}],"must":{"match_all":{}}}},"size":20,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
"
`;
